### PR TITLE
[SYCL][XPTI] Improved TLS feature for new SYCL instrumentation

### PR DESCRIPTION
**Pull request by tovinkere**
This PR improves tracepoint_t class that supports universal ID propagation through TLS. The documentation has been modified to reflect this change. The ability to stash code_location object in the TLS and capturing it in downstream functions to create a trace point has been added. This will enable us to instrument missing entry points without code location information to support user requests when it breaks ABI backward compatibility and easily enable full information during the ABI breakage window with the new features.

Summary:
 - Updated tracepoint_t class for Universal ID Propagation 
 - checkpoint_t class for publishing Universal ID through TLS, especially for the PI layer so Level 0/OpenCL etc can pick it up if needed for correlation. 
 - Only queue.copy() provides full code location information as it does not break ABI
 - queue(),  queue.memcpy() and queue.memset() instrumentation has been added without code location as it has been a frequest end-user request. The full code location information will be enabled in the next ABI breakage window.
 - USM allocations methods have been updated to notify (earlier, they only implemented the tracepoint)

### CODE REVIEW

**Comment by tovinkere** [comment]
Unfortunately, the failure is recursive. If I fix it with the patch provided, it wants me to revert back to previous format. Not sure how to address this.

**Comment by bader** [comment]
> Unfortunately, the failure is recursive. If I fix it with the patch provided, it wants me to revert back to previous format. Not sure how to address this.

I don't think the failure is recursive. There is a difference between clang-format suggested and applied changes.

File: sycl/source/detail/queue_impl.hpp
Comment (romanovvlad) [inline sycl/source/detail/queue_impl.hpp:106]: Could you please clarify why we need this commented out code here?

File: sycl/source/detail/queue_impl.hpp
Comment (romanovvlad) [inline sycl/source/detail/queue_impl.hpp:128]: It shouldn't be an ABI break if new member variables are added to the `queue_impl` class.

File: sycl/source/queue.cpp
Comment (romanovvlad) [inline sycl/source/queue.cpp:86]: This should be an ABI breaking change that are not allowed now.

**Comment by aelovikov-intel** [comment]
> This PR adds improves tracepoint_t class

Needs some re-wording as having two verbs looks wrong.

File: sycl/include/sycl/detail/common.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/common.hpp:131]: I'd reformat to start each numbered item in new line. Something like

```
some text... data in TLS:
  1) Populate ... text
     more text
      
  2) Remove ...
```

Also, s/when you at the top/when you are at the top/ ?

File: sycl/include/sycl/detail/common.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/common.hpp:123]: This needs overall description of what is being done. Why would we need a global variable for this? I expect I'll realize after reading the rest of this PR, but we need a short summary here for the readers after this is merged.

File: sycl/include/sycl/detail/common.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/common.hpp:147]: nit: I'd indent the code example by extra two spaces (I.e., everything except "usage" line).

File: sycl/include/sycl/detail/common.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/common.hpp:135]: Please modify the example to show what is expected to be user's code and what is SYCL RT and where is the boundary between them. It's still not clear at this moment what problem you're trying to solve.

File: sycl/source/detail/common.cpp
Comment (aelovikov-intel) [inline sycl/source/detail/common.cpp:18]: I guess this is the reasoning then. Do we expect to eliminate this during next ABI breaking window?

Also, I'm still not sure how it would work for the old binaries running with the new RT. What code location will we have in the TLS? How is it going to be useful?

File: sycl/source/detail/queue_impl.cpp
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.cpp:69]: Are we going to use this for all new code or only if tracing is being added for the existing code to avoid ABI-breakage? If the former, then PR should be split into two - infrastructure + rewrite of all existing codes and the 2nd PR containing this new instrumentatio (IIUC). If the latter, then extra comments on that are needed near the class definition.

File: sycl/source/detail/queue_impl.hpp
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.hpp:123]: IMO, this is way too verbose. I'd rather see something like

```
void xpti_notify(auto Callback) {
  #if XPTI_ENABLE_INSTRUMENTATION
  detail::tls_code_loc_t Tls;
  .... // rest of code here.
  Callback(getTEventFromTLS(TLS)); // To add callsite specific data
  ... // whatever common is needed after specific data
  #endif
}

void foo() {
  xpti_notify([](auto TEvent) { 
     xpti::addMetadata(TEvent, "src_memory_ptr", reinterpret_cast<size_t>(Src));
  });
}
```

**Review by aelovikov-intel — CHANGES_REQUESTED** [review]
I'm still not finished reading this but I'd like to pause and request for a high-level design (either a doc file or an elaborate high-level comment somewhere).

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:171]: This has been addressed in the new revision.

File: sycl/source/detail/queue_impl.cpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.cpp:67]: Using a combination of this suggestion and the one suggested by Andrei Elovikov.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp:106]: This was added as a guideline for future work. It has since been uncommented notifications provided without code location information as it is a frequent request from end users. The ability to separate code location information from the instrumentation allows us to make progress without breaking ABI compatibility. When ABI breakage window is available, the code location information can easily be enabled.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp:128]: Okay, I assumed class layout was affected - have updated it to cache it in the class now.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:131]: Updated the comment to reflect this request.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:123]: Done! - added more to the comments.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:147]: :-) Agreed.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:135]: Updated comments to make it a bit clearer.

File: sycl/source/detail/common.cpp
Comment (tovinkere) [reply sycl/source/detail/common.cpp:18]: SYCL will continue to grow over time and we will need to balance when to accommodate ABI breakage and when to provide end user notification. Code location information will make it complete, but with lack of visibility today, end users welcome notifications for missing functionality with partial information and augmenting them with code location like end user information when the ABI breakage window is available.

File: sycl/source/detail/queue_impl.cpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.cpp:69]: Existing functionality doesn't have to be rewritten as we propagate code location information all the way to the SYCL object.   How we approach this moving forward will change for all new instrumentation. With so many gaps currently in the instrumentation, I would like to provide end users with new data now and address refactoring existing instrumentation at a later time.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp:123]: Yes, thank you for the suggestion. I have updated the code to reflect this.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp]: Concise implementation checked in along similar lines.

File: sycl/source/detail/queue_impl.cpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.cpp]: Combining yours and Andrei's suggestions in the refactoring of the instrumentation.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp]: No longer commented as current features allow us to separate adding instrumentation features from adding code location information which usually breaks ABI. The instrumentation is now enabled and will augment with code location information in the next ABI breakage window as this is a requested instrumentation feature.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp]: Okay, caching the information in member variables now.

File: sycl/source/queue.cpp
Comment (tovinkere) [reply sycl/source/queue.cpp]: Reverted code location information, but going ahead with notifications as this is also a heavily requested instrumentation feature.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp]: Have address this with additional documentation that outlines what top of the stack means.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp]: Should have more documentation and the PR summary as been updated as well.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp]: Done!

File: sycl/source/detail/queue_impl.cpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.cpp]: No, existing instrumentation doesn't have to be changed. We could refactor it at a later time. Current PRs are to address requested features from end users that cover instrumentation gaps.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp]: Thanks for the suggestion - have used this outline from you and some suggestions from Vlad to refactor this to be more concise.

File: sycl/source/detail/queue_impl.hpp
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.hpp:201]: Can we make it a member function of a `TEvent` (or of a helper wrapper around it)? I'd like to see 

```
TEvent->addMetadata("sycl_device_name", MDevice->getDeviceName())
      ->addMetadata(...);
```

File: sycl/source/detail/queue_impl.hpp
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.hpp:184]: This looks *exactly* the same as the block at line 108. How can we eliminate code duplication?

File: sycl/source/detail/queue_impl.hpp
Comment (aelovikov-intel) [inline sycl/source/detail/queue_impl.hpp:679]:
```suggestion
  // All member variable defined here are needed for the SYCL instrumentation
```

File: sycl/source/detail/usm/usm_impl.cpp
Comment (aelovikov-intel) [inline sycl/source/detail/usm/usm_impl.cpp:47]: Are those cast absolutely necessary? If so, can we modify the ctor so that caller could avoid them (extra overload maybe?)

File: xpti/include/xpti/xpti_data_types.h
Comment (aelovikov-intel) [inline xpti/include/xpti/xpti_data_types.h:379]: Why isn't it an ABI break?

File: sycl/include/sycl/detail/common.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/common.hpp:138]: Assuming `bar` is at the ABI boundary, I am still missing how this is better than

```
// TODO: Remove at the next ABI break window.
void bar() [[deprecated]] {
  bar(code_location{});
}
void bar(const code_location &) {
 // Actual implementation.
}
```

So far, this is my biggest concern about this patch.

File: sycl/source/detail/xpti_registry.hpp
Comment (aelovikov-intel) [inline sycl/source/detail/xpti_registry.hpp:182]: It seems
```
notify()
scopedNotify()
```
always go in pair in this patch. Would that be universally so? If yes, can we call `notify` from `scopedNotify` and thus simplify users?

File: sycl/source/detail/xpti_registry.hpp
Comment (aelovikov-intel) [inline sycl/source/detail/xpti_registry.hpp:165]: That answers my concern about `(auto _&_ TEvent)`, although I'm not sure if I like that answer. I'd still prefer to introduce a helper wrapper class for that `TEvent` param (see my other comment about `addMetadata` in such a wrapper).

File: sycl/source/detail/xpti_registry.hpp
Comment (aelovikov-intel) [inline sycl/source/detail/xpti_registry.hpp:222]: I think...

```suggestion
  // The instance ID for the trace event; if it is called in a loop, then the
```

File: xpti/include/xpti/xpti_trace_framework.hpp
Comment (aelovikov-intel) [inline xpti/include/xpti/xpti_trace_framework.hpp:500]: nit: I'd like to see an early return instead.

File: xpti/include/xpti/xpti_trace_framework.hpp
Comment (aelovikov-intel) [inline xpti/include/xpti/xpti_trace_framework.hpp:625]: Likely a stupid question, but why is this in the same PR as the introduction of TLS? I would have expected we need mostly new code, not extensive modifications to the existing one.

File: xpti/include/xpti/xpti_trace_framework.hpp
Comment (aelovikov-intel) [inline xpti/include/xpti/xpti_trace_framework.hpp:705]: Why no `xptiTraceEnabled()` check?

**Review by aelovikov-intel — DISMISSED** [review]

File: sycl/source/detail/queue_impl.cpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.cpp:73]: Since it is a pointer to the object, we should be fine.

File: sycl/source/detail/common.cpp
Comment (tovinkere) [reply sycl/source/detail/common.cpp:35]: Whether we have code location or not, the notifications are valid. If we have a TLS from a higher scope, then the object will pick that up instead of the any local code location information.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:154]: Done!

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:170]: Done!

File: sycl/source/detail/common.cpp
Comment (tovinkere) [reply sycl/source/detail/common.cpp:33]: Done!

File: sycl/source/detail/queue_impl.cpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.cpp:66]: The IF block is enabled at compile time for building SYCL through SYCL_ENABLE_XPTI_TRACING. If this is set to off, one should not see any impact from XPTI.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp:201]: This is a good suggestion. There are similar discussions for refactoring XPTI, but current goals are to complete the tracing capability.

File: sycl/include/sycl/detail/common.hpp
Comment (tovinkere) [reply sycl/include/sycl/detail/common.hpp:138]: The initial implementation of tracing did this and we had to update the signature at every layer in the stack. The TLS approach allows us to bypass and also apply to SW layers beyond SYCL call chain.

File: sycl/source/detail/xpti_registry.hpp
Comment (tovinkere) [reply sycl/source/detail/xpti_registry.hpp:182]: Notify is to notify creation of an object in the asynchronous task graph. ScopedNotify() applies to action scopes. This allows any subscriber to the scoped notify trace type to measure duration times, event counts etc.

File: xpti/include/xpti/xpti_trace_framework.hpp
Comment (tovinkere) [reply xpti/include/xpti/xpti_trace_framework.hpp:500]: Done!

File: xpti/include/xpti/xpti_trace_framework.hpp
Comment (tovinkere) [reply xpti/include/xpti/xpti_trace_framework.hpp:625]: Initial TLS implementation committed over a year ago always assumed that the TLS contained a 64-bt universal ID that was to be used by the rest of the SW layers in the scope. However, the ability to stash code location was added to allow the notification to be decoupled from instrumenting the SYCL entry points. This allows to continue to close feature gaps without breaking ABI and enable code location information when ABI breakage is allowed.

File: xpti/include/xpti/xpti_trace_framework.hpp
Comment (tovinkere) [reply xpti/include/xpti/xpti_trace_framework.hpp:705]: An if check may be more expensive than the simple statement?

File: xpti/include/xpti/xpti_data_types.h
Comment (tovinkere) [reply xpti/include/xpti/xpti_data_types.h:379]: Good catch! As it was implemented, it is incorrect as scoped notify will fail. This change was attempting to fix that issue. Looks like we need to revert back until ABI breakage is allowed and handled it separately until then.

File: xpti/include/xpti/xpti_data_types.h
Comment (tovinkere) [reply xpti/include/xpti/xpti_data_types.h:402]: No, because the XPTI_TRACE_POINT_BEGIN/END  macros result in (25 <<1 | 0 ) and (25 << 1 | 1) respectively.

Comment (KseniyaTikhomirova) [reply xpti/include/xpti/xpti_data_types.h:402]: Ok, thanks

File: sycl/source/detail/queue_impl.cpp
Comment (aelovikov-intel) [reply sycl/source/detail/queue_impl.cpp:73]: `auto *` then?

**Comment by tovinkere** [comment]
@romanovvlad Any idea what the two failures are?

**Comment by romanovvlad** [comment]
> @romanovvlad Any idea what the two failures are?

No idea.

File: sycl/source/detail/queue_impl.hpp
Comment (tovinkere) [reply sycl/source/detail/queue_impl.hpp:688]: Removed the comment!

File: sycl/source/detail/usm/usm_impl.cpp
Comment (tovinkere) [reply sycl/source/detail/usm/usm_impl.cpp:134]: Thanks!

**Comment by tovinkere** [comment]
I am assuming that the failure can be ignored for now as it already has an issue opened against it: https://github.com/intel/llvm/issues/6495. 

Waiting on reviews from @smaslov-intel  and @intel/dpcpp-specification-reviewers - please prioritize.

**Review by smaslov-intel — APPROVED** [review]

**Comment by bader** [comment]
@tovinkere, could you address merge conflicts, please?

**Comment by tovinkere** [comment]
> > This PR improves tracepoint_t class that supports universal ID propagation through TLS. The documentation has been modified to reflect this change.
> 
> Are you referring to Doxygen documentation?
> 
> I see just one line change in sycl/doc/design/SYCLInstrumentationUsingXPTI.md. Should we document this patch improvements in the design document (i.e. sycl/doc/design/SYCLInstrumentationUsingXPTI.md)?

This is a reasonable request - I will see if an update is necessary for some of the samples. There are other suggestions by Andrei that I may incorporate into a future refactoring PR and a more comprehensive update to the design document may be appropriate at that time.

**Comment by bader** [comment]
@intel/llvm-reviewers-runtime, could you take a look, please?

**Comment by bader** [comment]
Resolved merge conflicts with https://github.com/intel/llvm/commit/18718fd842e96c2f3504b2b62f93cbc642350415.

**Comment by tovinkere** [comment]
Will this be merged soon? I thought @gmlueck had mentioned that a review by dpcpp-specification-reviewers wasn't needed.

**Pull request merged**
